### PR TITLE
[DATA PRODUCT] - Collection Summary Report

### PR DESCRIPTION
**Pull request by manjudr**

### CODE REVIEW

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  def getUserData(spark: SparkSession, fetchData: (SparkSession, Map[String, String], String, StructType) => DataFrame): DataFrame = {
+    val schema = Encoders.product[UserData].schema
+    fetchData(spark, userCacheDBSettings, "org.apache.spark.sql.redis", schema)
+      .withColumn("username", concat_ws(" ", col("firstname"), col("lastname"))).persist(StorageLevel.MEMORY_ONLY)
+  }
+
+  def getUserEnrollment(spark: SparkSession, fetchData: (SparkSession, Map[String, String], String, StructType) => DataFrame): DataFrame = {
+    fetchData(spark, userEnrolmentDBSettings, cassandraUrl, new StructType())
+      .select(col("batchid"), col("userid"), col("courseid"), col("active")
+        , col("completionpercentage"), col("enrolleddate"), col("completedon"), col("status"), col("certificates"), col("issued_certificates")).persist(StorageLevel.MEMORY_ONLY)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:91]: This should not be persisted

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  def prepareReport(spark: SparkSession, fetchData: (SparkSession, Map[String, String], String, StructType) => DataFrame, batchList: List[String])(implicit fc: FrameworkContext, config: JobConfig): DataFrame = {
+    implicit val sparkSession: SparkSession = spark
+    implicit val sqlContext: SQLContext = spark.sqlContext
+    import spark.implicits._
+    val userCachedDF = getUserData(spark, fetchData = fetchData).select("userid", "userchannel").persist(StorageLevel.MEMORY_ONLY)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:98]: This is persisted twice

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    val courseIds = processBatches.select(col("courseid")).distinct().collect().map(_ (0)).toList.asInstanceOf[List[String]]
+    val courseInfo = CourseUtils.getCourseInfo(courseIds, 500).toDF("framework", "identifier", "name", "channel", "batches", "organisation")
+    JobLogger.log(s"Total courseInfo records ${courseInfo.count()}", None, INFO)
+    val filteredBatches = processBatches.join(courseInfo, processBatches.col("courseid") === courseInfo.col("identifier"), "inner")
+      .select(processBatches.col("*"), courseInfo.col("identifier"), courseInfo.col("channel"), courseInfo.col("name"), courseInfo.col("organisation"))
+      .withColumn("batchid", concat(lit("batch-"), col("batchid")))
+      .withColumn("collectionName", col("name"))
+      .withColumn("publishedBy", concat_ws(", ", col("organisation")))
+      .withColumn("isPDFCertificatedIssued", when(col("certificates").isNotNull && size(col("certificates").cast("array<map<string, string>>")) > 0, true).otherwise(false))
+      .withColumn("isSVGCertificatedIssued", when(col("issued_certificates").isNotNull && size(col("issued_certificates").cast("array<map<string, string>>")) > 0, true).otherwise(false))
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:112]: Refer to the progressexhaust job logic and convert them into isCertificateIssued flag directly

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+      .withColumn("isCertified", when((col("isPDFCertificatedIssued") === true || col("isSVGCertificatedIssued") === true), "Y").otherwise("N"))
+      .withColumn("isSameOrgUserEnrolled", when((col("userchannel") === col("channel")), true).otherwise(false))
+      .withColumn("isSameOrgUserCompleted", when((col("userchannel") === col("channel") && col("status") === 2), true).otherwise(false))
+      .withColumn("enrolDate", to_timestamp(col("enrolleddate"), fmt = "yyyy-MM-dd HH:mm:ss"))
+      .withColumn("completedDate", to_timestamp(col("completedon"), fmt = "yyyy-MM-dd HH:mm:ss"))
+      .withColumn("diffInMinutes", round(col("completedDate").cast(LongType) - col("enrolDate").cast(LongType)) / 60) // Converting to mins
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:118]: Use a UDF and get the elapsedTime directly.

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      .drop("isPDFCertificatedIssued", "isSVGCertificatedIssued").persist(StorageLevel.MEMORY_ONLY)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:119]: Not required to persist

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+      .filter(col("isSameOrgUserCompleted") === true)
+    val computedDF = transformedDF.dropDuplicates("courseid", "batchid").join(totalEnrolledUsers, Seq("courseid", "batchid"), "left_outer")
+      .join(completedUserCount, Seq("courseid", "batchid", "status"), "left_outer")
+      .join(certificates, Seq("courseid", "batchid", "isCertified"), "left_outer")
+      .join(sameOrgEnrolledUserCount, Seq("courseid", "batchid", "isSameOrgUserEnrolled"), "left_outer")
+      .join(sameOrgCompletedUserCount, Seq("courseid", "batchid", "isSameOrgUserCompleted"), "left_outer")
+      .withColumn("avgElapsedTime", transformedDF.col("diffInMinutes") / col("completionUserCount"))
+      .select("batchid", "courseid", "collectionName", "publishedBy", "startdate", "enddate", "enrolledUsersCount", "completionUserCount", "sameOrgEnrolledUserCount", "sameOrgCompletedUserCount", "certificatedIssuedCount", "isCertified", "avgElapsedTime")
+    computedDF
+  }
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:145]: Why are there so many grouped queries? do one groupBy and multiple aggregates

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  override def main(config: String)(implicit sc: Option[SparkContext] = None, fc: Option[FrameworkContext] = None) {
+    JobLogger.init(jobName)
+    JobLogger.start(s"$jobName started executing", Option(Map("config" -> config, "model" -> jobName)))
+    implicit val jobConfig: JobConfig = JSONUtils.deserialize[JobConfig](config)
+    implicit val spark: SparkSession = openSparkSession(jobConfig)
+    implicit val frameworkContext: FrameworkContext = getReportingFrameworkContext()
+    init()
+    val conf = config.split(";")
+    val batchIds = if (conf.length > 1) {
+      conf(1).split(",").toList
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:66]: config should be similar to exhaust jobs. We should have capability to run for batches matching a search query

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+      JobLogger.end(s"$jobName completed execution", "SUCCESS", Option(Map("timeTaken" -> res._1, "totalRecords" -> res._2.count())))
+    } finally {
+      frameworkContext.closeContext()
+      spark.close()
+    }
+  }
+
+  // $COVERAGE-OFF$ Disabling scoverage for main and execute method
+  def init()(implicit spark: SparkSession, fc: FrameworkContext, config: JobConfig) {
+    DecryptUtil.initialise()
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:80]: We don't need this

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    fetchData(spark, userCacheDBSettings, "org.apache.spark.sql.redis", schema)
+      .withColumn("username", concat_ws(" ", col("firstname"), col("lastname")))
+      .withColumn("userstate", when(col("state").isNotNull && col("state") =!= "", col("state")).otherwise("NA"))
+  }
+
+  def getUserEnrollment(spark: SparkSession, fetchData: (SparkSession, Map[String, String], String, StructType) => DataFrame): DataFrame = {
+    fetchData(spark, userEnrolmentDBSettings, cassandraUrl, new StructType())
+      .select(col("batchid"), col("userid"), col("courseid"), col("active")
+        , col("completionpercentage"), col("enrolleddate"), col("completedon"), col("status"), col("certificates"), col("issued_certificates"))
+  }
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:98]: We should filter for active users only

Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:98]: Why are we selecting "active" and "completionpercentage" columns if we are going to drop them off?

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    import spark.implicits._
+    val userCachedDF = getUserData(spark, fetchData = fetchData).select("userid", "userchannel", "userstate")
+    val processBatches: DataFrame = filterBatches(spark, fetchData, config, batchList)
+      .join(getUserEnrollment(spark, fetchData), Seq("batchid", "courseid"), "left_outer")
+      .join(userCachedDF, Seq("userid"), "left_outer").drop("completionpercentage", "active")
+
+    val courseIds = processBatches.select(col("courseid")).distinct().collect().map(_ (0)).toList.asInstanceOf[List[String]]
+    val courseInfo = CourseUtils.getCourseInfo(courseIds, 500).toDF("framework", "identifier", "name", "channel", "batches", "organisation")
+    JobLogger.log(s"Total courseInfo records ${courseInfo.count()}", None, INFO)
+    val filteredBatches = processBatches.join(courseInfo, processBatches.col("courseid") === courseInfo.col("identifier"), "inner")
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:113]: Join the course data once the aggregates are computed

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+      .withColumn("hasCertified", when(col("certificateIssueCount") > 0, "Y").otherwise("N"))
+      .select("publishedBy", "batchid", "courseid", "collectionName", "startdate", "enddate", "hasCertified", "userstate", "enrolledUsersCountByState", "completionUserCountByState", "certificateIssueCount")
+  }
+
+  def saveToBlob(reportData: DataFrame): Unit = {
+    val container = AppConf.getConfig("cloud.container.reports")
+    val objectKey = AppConf.getConfig("course.metrics.cloud.objectKey")
+    val storageConfig = getStorageConfig(container, objectKey)
+    JobLogger.log(s"Uploading reports to blob storage", None, INFO)
+    val reportPath = "collection-summary-reports/"
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:140]: Don't hardcode the path

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    val fields = reportData.schema.fieldNames
+    val colNames = for (e <- fields) yield columnMapping.getOrElse(e, e)
+    val dynamicColumns = fields.toList.filter(e => !columnMapping.keySet.contains(e))
+    val columnWithOrder = (columnsOrder ::: dynamicColumns).distinct
+    reportData.toDF(colNames: _*).select(columnWithOrder.head, columnWithOrder.tail: _*)
+      .saveToBlobStore(storageConfig, "csv", s"${reportPath}summary-report-${getDate}", Option(Map("header" -> "true")), None)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:146]: File name should always be constant. Create a file with <report-id>_<date>.csv and <report-id>_latest.csv

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    implicit val sqlContext: SQLContext = spark.sqlContext
+    import spark.implicits._
+    val userCachedDF = getUserData(spark, fetchData = fetchData).select("userid", "userchannel", "userstate")
+    val processBatches: DataFrame = filterBatches(spark, fetchData, config)
+      .join(getUserEnrollment(spark, fetchData), Seq("batchid", "courseid"), "left_outer")
+      .join(userCachedDF, Seq("userid"), "left_outer")
+
+    val filteredBatches = processBatches.select(col("*"))
+      .withColumn("isCertified", when(col("certificates").isNotNull && size(col("certificates").cast("array<map<string, string>>")) > 0, "Y")
+        .when(col("issued_certificates").isNotNull && size(col("issued_certificates").cast("array<map<string, string>>")) > 0, "Y").otherwise("N"))
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:106]: Move this logic to the enrollment DF itself

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+      val courseIds = processedBatches.select(col("courseid")).distinct().collect().map(_ (0)).toList.asInstanceOf[List[String]]
+      val courseInfo = CourseUtils.getCourseInfo(courseIds, None, config.modelParams.get.getOrElse("maxlimit", 500).asInstanceOf[Int]).toDF("framework", "identifier", "name", "channel", "batches", "organisation")
+      JobLogger.log(s"Total courseInfo records ${courseInfo.count()}", None, INFO)
+      processedBatches.join(courseInfo, processedBatches.col("courseid") === courseInfo.col("identifier"), "inner")
+        .withColumn("collectionName", col("name"))
+        .withColumn("publishedBy", concat_ws(", ", col("organisation")))
+    } else {
+      processedBatches
+    }
+    reportDF.select("publishedBy", "batchid", "courseid", "collectionName", "startdate", "enddate", "hasCertified", "userstate", "enrolledUsersCountByState", "completionUserCountByState", "certificateIssueCount")
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:119]: Move this into filteredCols similar to exhaust jobs

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+  def computeValues(transformedDF: DataFrame): DataFrame = {
+    // Compute completionCount and enrolCount for state
+    val statePartitionDF = transformedDF.groupBy("batchid", "courseid", "userstate").agg(
+      count(when(col("status") === 2, 1)).as("completionUserCountByState"),
+      count(when(col("isCertified") === "Y", 1)).as("certificateIssueCount"),
+      count(col("userid")).as("enrolledUsersCountByState")
+    )
+    statePartitionDF.join(transformedDF.drop("isCertified", "status", "userstate").dropDuplicates("courseid", "batchid"), Seq("courseid", "batchid"), "inner")
+      .withColumn("batchid", concat(lit("batch-"), col("batchid")))
+      .withColumn("hasCertified", when(col("certificateIssueCount") > 0, "Y").otherwise("N"))
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:131]: Shouldn't this come from content model?

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+  }
+
+  def computeValues(transformedDF: DataFrame): DataFrame = {
+    // Compute completionCount and enrolCount for state
+    val statePartitionDF = transformedDF.groupBy("batchid", "courseid", "userstate").agg(
+      count(when(col("status") === 2, 1)).as("completionUserCountByState"),
+      count(when(col("isCertified") === "Y", 1)).as("certificateIssueCount"),
+      count(col("userid")).as("enrolledUsersCountByState")
+    )
+    statePartitionDF.join(transformedDF.drop("isCertified", "status", "userstate").dropDuplicates("courseid", "batchid"), Seq("courseid", "batchid"), "inner")
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:129]: Why are we joining back with transformedDF?

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala (hunk 15 above)
Comment (manjudr) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:131]: I have checked with Pradyumna, As of now in the content model, they are not storing any certificates related information.

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala (hunk 16 above)
Comment (manjudr) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala:129]: The statePartitionDF will be having only "(batchid, courseid, certificateIssueCount, hasCertified,  certificateIssueCount, enrolmentCount and state)"
But the startdate and enddate are will be in the "transformedDF" so I'm doing "inner" with that.

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala (hunk 15 above)
Comment (manjudr) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/CollectionSummaryJob.scala]: Marking the flag by referring course_batch table

**Review by SanthoshVasabhaktula — APPROVED** [review]
Test case assertions need to be added

**Pull request merged**